flight/Modules/CharacterOSD/panel.c

```diff
@@ -0,0 +1,471 @@
+		mode = "ST2";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_STABILIZED3:
+		mode = "ST3";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_AUTOTUNE:
+		mode = "TUNE";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_ALTITUDEHOLD:
+		mode = "AHLD";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_POSITIONHOLD:
+		mode = "PHLD";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_RETURNTOHOME:
+		mode = "RTH";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_PATHPLANNER:
+		mode = "PLAN";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_FAILSAFE:
+		mode = "FAIL";
+		break;
+	case FLIGHTSTATUS_FLIGHTMODE_TABLETCONTROL:
+		// There are many sub modes here that could be filled in.
+		mode = "TAB";
+		break;
+	}
+
+	draw_rect(state, x, y, 6, 3, 0, 0);
+	PIOS_MAX7456_puts(state->dev, x + 1, y + 1, mode, 0);
+}
+
+/* ArmedFlag */
+static void ARMEDFLAG_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	uint8_t attr = state->telemetry.flight_status.arm_status == FLIGHTSTATUS_ARMED_ARMED ? 0 : MAX7456_ATTR_INVERT;
+	draw_rect(state, x, y, 3, 3, true, attr);
+	PIOS_MAX7456_put(state->dev, x + 1, y + 1, 0xe0, attr);
+}
+
+static void FLIGHTTIME_update(charosd_state_t state, uint8_t x, uint8_t y) {
+       char buffer[10];
+       uint32_t time = state->telemetry.system.flight_time;
+       int min, sec;
+
+       uint16_t hours = (time / 3600000); // hours
+       if (hours == 0) {
+               min = time / 60000;
+               sec = (time / 1000) - 60 * min;
+               sprintf(buffer, "%02d:%02d", (int)min, (int)sec);
+       } else {
+               min = time / 60000 - 60 * hours;
+               sec = (time / 1000) - 60 * min - 3600 * hours;
+               sprintf(buffer, "%02d:%02d:%02d", (int)hours, (int)min, (int)sec);
+       }
+
+       terminate_buffer();
+
+       PIOS_MAX7456_puts(state->dev, x + 1, y + 1, buffer, 0);
+}
+
+/* Roll */
+STD_PANEL(ROLL, 7, "\xb2%d\xb0", (int16_t) state->telemetry.attitude_actual.Roll);
+
+/* Pitch */
+STD_PANEL(PITCH, 7, "\xb1%d\xb0", (int16_t) state->telemetry.attitude_actual.Pitch);
+
+/* GPS */
+
+#define _PAN_GPS_2D 0x01
+#define _PAN_GPS_3D 0x02
+
+static void GPS_update(charosd_state_t state, uint8_t x, uint8_t y)
+{
+	char buffer[4];
+
+	snprintf(buffer, sizeof(buffer), "%d", state->telemetry.gps_position.Satellites); 
+	terminate_buffer();
+	bool err = state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX2D;
+	PIOS_MAX7456_puts(state->dev, x, y, "\x10\x11", err ? MAX7456_ATTR_INVERT : 0);
+	PIOS_MAX7456_put(state->dev, x + 2, y, state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX3D ? _PAN_GPS_2D : _PAN_GPS_3D,
+		err ? MAX7456_ATTR_INVERT : (state->telemetry.gps_position.Status < GPSPOSITION_STATUS_FIX2D ? MAX7456_ATTR_BLINK : 0));
+	if (err) PIOS_MAX7456_puts (state->dev, x + 3, y, ERR_STR, MAX7456_ATTR_INVERT);
+	else PIOS_MAX7456_puts(state->dev, x + 3, y, buffer, 0);
+}
+
+/* Lat */
+STD_PANEL(LATITUDE, 11, "\x83%02.6f", (double)state->telemetry.gps_position.Latitude / 10000000.0);
```

Comment:
ugh.  These are marginal/ should probably remain as doubles.  Or we could do fixed point explicitly (carefully)